src/_cffi_src/openssl/src/osrandom_engine.c

```diff
@@ -73,9 +84,22 @@ static int osrandom_init(ENGINE *e) {
 
 static int osrandom_rand_bytes(unsigned char *buffer, int size) {
     ssize_t n;
+    struct stat st;
+
+    if (fstat(urandom_cache.fd, &st)
+        || st.st_dev != urandom_cache.st_dev
+        || st.st_ino != urandom_cache.st_ino) {
+        /* Something changed: forget the cached fd (but don't close it,
+           since it probably points to something important for some
+           third-party code). Re-initialize the engine. */
+        int res = osrandom_init(NULL);
```

Comment:
That sounds reasonable.

On Wed, Jun 17, 2015 at 11:35 AM, Paul Kehrer <notifications@github.com>
wrote:

> In src/_cffi_src/openssl/src/osrandom_engine.c
> <https://github.com/pyca/cryptography/pull/2035#discussion_r32639941>:
>
> > @@ -73,9 +84,22 @@ static int osrandom_init(ENGINE *e) {
> >
> >  static int osrandom_rand_bytes(unsigned char *buffer, int size) {
> >      ssize_t n;
> > +    struct stat st;
> > +
> > +    if (fstat(urandom_cache.fd, &st)
> > +        || st.st_dev != urandom_cache.st_dev
> > +        || st.st_ino != urandom_cache.st_ino) {
> > +        /* Something changed: forget the cached fd (but don't close it,
> > +           since it probably points to something important for some
> > +           third-party code). Re-initialize the engine. */
> > +        int res = osrandom_init(NULL);
>
> I'd be okay with erroring out here rather than trying to recover. We could
> also push an error onto the OpenSSL stack so we might have a chance of
> figuring out why this is happening in the future when some bad software
> closes all fds on us.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pyca/cryptography/pull/2035/files#r32639941>.
>



-- 
"I disapprove of what you say, but I will defend to the death your right to
say it." -- Evelyn Beatrice Hall (summarizing Voltaire)
"The people's good is the highest law." -- Cicero
GPG Key fingerprint: 125F 5C67 DFE9 4084